pkg/detector/detector.go

```diff
@@ -615,49 +630,34 @@ func (d *ResourceDetector) GetUnstructuredObject(objectKey keys.ClusterWideKey)
 }
 
 // ClaimPolicyForObject set policy identifier which the object associated with.
-func (d *ResourceDetector) ClaimPolicyForObject(object *unstructured.Unstructured, policyNamespace, policyName, policyUID string) error {
+func (d *ResourceDetector) ClaimPolicyForObject(object *unstructured.Unstructured, policy *policyv1alpha1.PropagationPolicy) error {
 	objLabels := object.GetLabels()
-	if objLabels == nil {
-		objLabels = make(map[string]string)
-	} else if len(objLabels) > 0 {
+	if len(objLabels) > 0 {
 		// object has been claimed, don't need to claim again
 		if !excludeClusterPolicy(objLabels) &&
-			objLabels[policyv1alpha1.PropagationPolicyNamespaceLabel] == policyNamespace &&
-			objLabels[policyv1alpha1.PropagationPolicyNameLabel] == policyName {
+			objLabels[policyv1alpha1.PropagationPolicyUIDLabel] == string(policy.UID) {
 			return nil
 		}
 	}
 
-	objLabels[policyv1alpha1.PropagationPolicyNamespaceLabel] = policyNamespace
-	objLabels[policyv1alpha1.PropagationPolicyNameLabel] = policyName
-	objLabels[policyv1alpha1.PropagationPolicyUIDLabel] = policyUID
-
-	objectAnnotations := object.GetAnnotations()
-	if objectAnnotations == nil {
-		objectAnnotations = make(map[string]string)
-	}
-	objectAnnotations[policyv1alpha1.PropagationPolicyNamespaceAnnotation] = policyNamespace
-	objectAnnotations[policyv1alpha1.PropagationPolicyNameAnnotation] = policyName
-
 	objectCopy := object.DeepCopy()
-	objectCopy.SetLabels(objLabels)
-	objectCopy.SetAnnotations(objectAnnotations)
+	util.MergeAnnotation(objectCopy, policyv1alpha1.PropagationPolicyNamespaceAnnotation, policy.Namespace)
```

Comment:
could you extend `util.MergeAnnotation` and `util.MergeLabel`? they are not efficient, every annotation and label need to get object's and set. they should looks like this:
```go
// MergeAnnotations adds annotations for the given object, keep the value unchanged if key exist.
func MergeAnnotations(obj *unstructured.Unstructured, annos map[string]string) {
	objectAnnotation := obj.GetAnnotations()
	if objectAnnotation == nil {
		objectAnnotation = make(map[string]string, len(annos))
	}

	for annotationKey, annotationValue := range annos {
		if _, exist := objectAnnotation[annotationKey]; !exist {
			objectAnnotation[annotationKey] = annotationValue
		}
	}
	obj.SetAnnotations(objectAnnotation)
}
```

Reply:
I think we can do it in another PR